File: mobile-apk/lib/services/auth_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';
import 'supabase_service.dart';
import '../security/input_validator.dart';

class AuthService {
  static String normalizeEmail(String value) => InputValidator.email(value);

  static Future<AuthResponse> signIn(String email, String password) =>
      SupabaseService.client.auth.signInWithPassword(
          email: normalizeEmail(email),
          password: InputValidator.password(password));

  static Future<AuthResponse> signUp(
      String email, String password, String fullName, String phone) async {
    final response = await SupabaseService.client.auth.signUp(
      email: normalizeEmail(email),
      password: InputValidator.password(password),
      data: {
        'full_name': InputValidator.fullName(fullName),
        'phone': InputValidator.phone(phone)
      },
    );
    if (response.user?.identities?.isEmpty == true)
      throw const AuthException('Email sudah terdaftar. Silakan masuk.');
    return response;
  }

  static Future<void> signOut() => SupabaseService.client.auth.signOut();
  static bool get isLoggedIn => SupabaseService.client.auth.currentUser != null;
}

File: frontend-web/src/services/authService.js
import { supabase } from "../supabase/client.js";

export async function signUp({ email, password, fullName, phone, options = {} }) {
  const metadata = {
    ...(options.data || {}),
    full_name: fullName,
    phone,
    account_type: "user",
  };
  const { data, error } = await supabase.auth.signUp({ email, password, options: { ...options, data: metadata } });
  if (error) throw error;
  if (data.user?.identities?.length === 0) throw new Error("Email sudah terdaftar. Silakan masuk.");
  return data;
}

export async function signIn({ email, password }) {
  const { data, error } = await supabase.auth.signInWithPassword({ email, password });
  if (error) throw error;
  return data;
}

export async function signInWithGoogle() {
  if (window.DimsumLumerApp?.postMessage) {
    window.DimsumLumerApp.postMessage(JSON.stringify({ type: "google_sign_in" }));
    return { native: true };
  }
  const redirectTo = new URL("/profil", window.location.origin).toString();
  const { data, error } = await supabase.auth.signInWithOAuth({
    provider: "google",
    options: { redirectTo },
  });
  if (error) throw error;
  return data;
}

// Dipanggil shell APK setelah OAuth native berhasil. Session tetap divalidasi
// oleh Supabase; frontend tidak menerima atau menyimpan kredensial Google.
window.addEventListener("dimsum-lumer-auth", async (event) => {
  const accessToken = event?.detail?.accessToken;
  const refreshToken = event?.detail?.refreshToken;
  if (!accessToken || !refreshToken) return;
  const { error } = await supabase.auth.setSession({
    access_token: accessToken,
    refresh_token: refreshToken,
  });
  if (!error) window.location.replace("/profil?source=apk");
});

export async function requestPasswordReset(email) {
  const redirectTo = new URL("/reset-password", window.location.origin).toString();
  const { data, error } = await supabase.auth.resetPasswordForEmail(email, { redirectTo });
  if (error) throw error;
  return data;
}

export async function updatePassword(password) {
  const { data,error }=await supabase.auth.updateUser({password});
  if(error)throw error;
  return data;
}

export async function verifyPasswordRecoveryOtp(email,token) {
  const {data,error}=await supabase.auth.verifyOtp({email:email.trim().toLowerCase(),token:token.trim(),type:"recovery"});
  if(error)throw error;
  return data;
}

export async function signOut() { const { error }=await supabase.auth.signOut();if(error)throw error; }
export async function getCurrentSession() { const {data,error}=await supabase.auth.getSession();if(error)throw error;return data.session; }
export const onAuthStateChange=(callback)=>supabase.auth.onAuthStateChange((event,session)=>callback(event,session));
